CHANGELOG.md

```diff
@@ -1,6 +1,136 @@
+      else -> // generic error
+    }
+  }
+}
+```
+
+To ease the migration to v4, the v3 behavior can be restored by calling `ApolloClient.Builder.useV3ExceptionHandling(true)`.
+
+Feedback about this change is welcome on [issue 4711](https://github.com/apollographql/apollo-kotlin/issues/4711).
+
+## ☕️ Better Java support
+
+As v3 has a Kotlin and Coroutines first API, using it from Java is sometimes impractical or not idiomatic. That is why in v4 we are introducing a new Java runtime, written in Java, which provides a Java friendly API. It is callback based and doesn't depend on a third-party library like Rx.
+
+To use it in your project, instead of the usual runtime (`com.apollographql.apollo3:apollo-runtime`), use the following dependency in your `build.gradle[.kts]` file:
+
+```kotlin
+implementation("com.apollographql.apollo3:apollo-runtime-java")
+```
+
+Then you can use the `ApolloClient` class from Java:
+
+```java
+ApolloClient apolloClient = new ApolloClient.Builder()
+  .serverUrl("https://...")
+  .build();
+
+apolloClient
+  .query(MyQuery.builder().build())
+  .enqueue(new ApolloCallback<MyQuery.Data>() {
+      @Override public void onResponse(@NotNull ApolloResponse<MyQuery.Data> response) {
+        System.out.prinitln(response.getData());
+      }
+  });
+```
+
+A few examples can be found in the [tests](https://github.com/apollographql/apollo-kotlin/tree/main/tests/java-client/src/test/java/test).
+
+## 🔃 Multi-module improvements
+
+In multi-module projects, by default, all the types of an upstream module are generated because there is no way to know in advance what types are going to be used by downstream modules. For large projects this can lead to a lot of unused code and an increased build time.
+
+To avoid this, in v3 you could manually specify which types to generate by using `alwaysGenerateTypesMatching`. In v4 this can now be computed automatically by detecting which types are used by the downstream modules.
+
+To enable this, add the "opposite" link of dependencies with `isADependencyOf()`.
+
+```kotlin
+// schema/build.gradle.kts
+apollo {
+  service("service") {
+    packageName.set("schema")
+
+    // Enable generation of metadata for use by downstream modules 
+    generateApolloMetadata.set(true)
+
+    // More options...
+
+    // Get used types from the downstream module
+    isADependencyOf(project(":feature1"))
+
+    // You can have several downstream modules
+    isADependencyOf(project(":feature2"))
+  }
+}
+```
+
+```kotlin
+// feature1/build.gradle.kts
+apollo {
+  service("service") {
+    packageName.set("feature1")
+    
+    // Get the generated schema types (and fragments) from the upstream schema module 
+    dependsOn(project(":schema")) 
+  }
+}
+```
+
+## ⚙️ Other changes
+
+- All the previously deprecated APIs have been removed.
+- For a more exhaustive list of changes, see the [migration guide](https://www.apollographql.com/docs/kotlin/v4/migration/4.0).
```

Comment:
That's a bit confusing as the migration guide doesn't really contain all the changes (doesn't contain new features for an example). All in all, I think I would remove the "other changes" section 


About migration guide, might be worth doing a quick path on it? Add the multi-module stuff + it contains some details about HTTP headers that I'm not 100% sure are still valid